## Task 5: Credit Card Fraud Detection

## Steps followed:
##### 1. Importing Libraries
##### 2. Importing Dataset
##### 3. Data Pre-processing
##### 4. Model Building and Training
##### 5. Model Testing and Evaluation

#### 1. Importing Libraries

In [44]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as mpl
from sklearn.linear_model import LogisticRegression as LR
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder 
from sklearn.metrics import accuracy_score, precision_score

#### 2. Importing Dataset

In [3]:
data = pd.read_csv('creditcard.csv')

In [4]:
data

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

#### 3. Data Pre-processing

In [5]:
data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [6]:
data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [8]:
data.shape

(284807, 31)

In [9]:
data.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

In [10]:
data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [13]:
data['Class'].unique()

array([0, 1], dtype=int64)

In the 'Class' column, value 0 represents legit transactions and value 1 represents fraudulent transactions. 

In [16]:
data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

We can observe that there are 492 fraudulent transactions. Therefore, it is an unbalanced dataset. 

We create separate data frames for normal and fraudulent transactions

In [17]:
normal_transactions = data[data.Class==0]
fraud_transactions  = data[data.Class==1]

In [18]:
print(normal_transactions.shape)
print(fraud_transactions.shape)

(284315, 31)
(492, 31)


In [19]:
normal_transactions['Amount'].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [20]:
fraud_transactions['Amount'].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [24]:
#Comparing values for both transactions
data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

This is a case of under-sampling. 
To solve this, we build a sample dataset containing similar distribution of both the classes. 

Number of fraudulent transactions: 492

Number of legit transactions:284315

We select 492 random transactions from the normal transactions dataset. 

In [25]:
normal_sample = normal_transactions.sample(n=492)

In [27]:
#Concatenating normal_sample and fraud_transactions to create one dataset. 
dataset=pd.concat([normal_sample,fraud_transactions],axis=0)
dataset

Time        V1        V2         V3        V4        V5        V6  \
115755   74001.0 -0.799734  0.736136   2.833526  0.974067 -0.514931  0.792281   
176170  122654.0 -0.231824  0.978727   0.895745  2.659376 -0.298415  1.703268   
12030    20823.0 -5.373899  5.537306 -10.451813  5.831008 -5.209512 -3.658238   
159933  113042.0  2.082490 -0.062589  -1.468650  0.183888  0.165087 -0.923195   
147602   88732.0 -0.237039  0.803321  -1.487158 -0.294079  3.241715  3.588285   
...          ...       ...       ...        ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653  -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381  -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366  -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864  -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476  -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
115755 -0.370207  0.557209 -0.299558  ...  0.346323  1.113320 -0.278916   
176170  1.738049 -0.287691 -0.751416  ...  0.128737  0.738265 -0.220916   
12030  -9.526906  3.996788 -3.224948  ...  1.550854 -0.446762 -0.080766   
159933  0.132132 -0.265853  0.826432  ...  0.257485  0.866644  0.016740   
147602  0.570079  0.602009 -0.168856  ...  0.303300  1.265967 -0.040283   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
115755  0.032092  0.103587 -0.026899  0.391301  0.149521   13.99      0  
176170 -1.000062 -0.314607  0.014293 -0.177144 -0.264249  274.29      0  
12030  -0.019581  0.104116 -0.304290  1.954876  0.686027    1.00      0  
159933  0.743621  0.338951 -0.451246  0.000420 -0.048152    1.00      0  
147602  0.715588 -0.751401 -0.451896  0.384764  0.192978   23.08      0  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[984 rows x 31 columns]

In [28]:
dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [29]:
dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93297.760163 -0.091056  0.030297  0.016966  0.047490 -0.065110   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.025380 -0.062017 -0.001887 -0.041212  ... -0.012291 -0.046102   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.019554 -0.021208 -0.021081 -0.037232 -0.003059 -0.022865  0.010846   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       91.113557  
1      122.211321  

[2 rows x 30 columns]

### 4. Model Building and Training

##### Feature Extraction

In [31]:
# X contains features i.e. all columns other than Class
# Y contains target variable i.e. Class

X=dataset.drop(columns='Class',axis=1)
Y=dataset['Class']

In [34]:
X

Time        V1        V2         V3        V4        V5        V6  \
115755   74001.0 -0.799734  0.736136   2.833526  0.974067 -0.514931  0.792281   
176170  122654.0 -0.231824  0.978727   0.895745  2.659376 -0.298415  1.703268   
12030    20823.0 -5.373899  5.537306 -10.451813  5.831008 -5.209512 -3.658238   
159933  113042.0  2.082490 -0.062589  -1.468650  0.183888  0.165087 -0.923195   
147602   88732.0 -0.237039  0.803321  -1.487158 -0.294079  3.241715  3.588285   
...          ...       ...       ...        ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653  -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381  -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366  -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864  -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476  -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
115755 -0.370207  0.557209 -0.299558  ...  0.243629  0.346323  1.113320   
176170  1.738049 -0.287691 -0.751416  ... -0.276246  0.128737  0.738265   
12030  -9.526906  3.996788 -3.224948  ...  1.461270  1.550854 -0.446762   
159933  0.132132 -0.265853  0.826432  ... -0.256938  0.257485  0.866644   
147602  0.570079  0.602009 -0.168856  ...  0.262926  0.303300  1.265967   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
115755 -0.278916  0.032092  0.103587 -0.026899  0.391301  0.149521   13.99  
176170 -0.220916 -1.000062 -0.314607  0.014293 -0.177144 -0.264249  274.29  
12030  -0.080766 -0.019581  0.104116 -0.304290  1.954876  0.686027    1.00  
159933  0.016740  0.743621  0.338951 -0.451246  0.000420 -0.048152    1.00  
147602 -0.040283  0.715588 -0.751401 -0.451896  0.384764  0.192978   23.08  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

In [35]:
Y

115755    0
176170    0
12030     0
159933    0
147602    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64

##### Splitting the dataset into training and testing models

In [36]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=2,stratify=Y)

In [37]:
print(f"X: {X.shape}, X_train:{X_train.shape},X_test:{X_test.shape}")
print(f"Y: {Y.shape}, Y_train:{Y_train.shape},Y_test:{Y_test.shape}")

X: (984, 30), X_train:(787, 30),X_test:(197, 30)
Y: (984,), Y_train:(787,),Y_test:(197,)


In [38]:
X_train.head()

Time        V1        V2        V3        V4        V5        V6  \
100623   67571.0 -0.758469 -0.045410 -0.168438 -1.313275 -1.901763  0.739433   
18371    29427.0 -0.501597  0.422965  1.380465 -1.813467 -0.020195 -1.564136   
222086  142815.0 -0.616383 -0.510969  0.251073 -2.186422 -0.172318  1.891172   
132511   80014.0 -1.963111 -0.951415  3.074674  3.305077  0.891820 -0.766261   
87919    61904.0  1.108159  0.541513  0.326021  2.713153  0.094432  0.196867   

              V7        V8        V9  ...       V20       V21       V22  \
100623  3.071892 -0.483422  0.618203  ... -0.032500  0.042619  0.397224   
18371   1.003791 -0.604658 -1.538044  ...  0.053460  0.245540  0.693395   
222086 -0.702285  1.043801 -0.923626  ... -0.259129 -0.232357 -0.448965   
132511 -1.564663  0.545652 -0.539545  ...  0.798389  0.403907  0.625782   
87919  -0.084217  0.143319 -0.493787  ... -0.093060 -0.103995 -0.151165   

             V23       V24       V25       V26       V27       V28  Amount  
100623  0.072229 -0.242276  0.560916 -0.540955  0.150606 -0.117140  549.06  
18371  -0.393209  0.740193  0.435795 -0.303265 -0.267074 -0.145918   12.00  
222086  0.500647 -0.983626 -1.335729  0.512410  0.080231  0.098417  105.88  
132511  0.164092  0.347772  0.060952  0.417551  0.290613  0.204409   30.27  
87919  -0.161348 -0.069333  0.639878  0.079552  0.016461  0.033675   15.21  

[5 rows x 30 columns]

In [39]:
X_test.head()

Time         V1         V2         V3        V4         V5  \
247995  153761.0   1.146259   1.403458  -4.159148  2.660107  -0.323217   
189888  128599.0  -5.738618  -3.473721   1.075626 -2.532636   0.113257   
15648    27060.0  -8.666683  -6.210269   1.797420  2.297472   0.825676   
165842  117701.0  -0.946224   0.437470   0.668355 -2.011183  -0.261253   
15451    26833.0 -20.532751  12.373989 -23.009003  6.144821 -15.587296   

              V6         V7         V8        V9  ...       V20       V21  \
247995 -1.836071  -1.623740   0.259562 -1.132044  ...  0.284831  0.564450   
189888 -0.588758  -1.661158   0.828246  3.707288  ... -1.609226 -0.368006   
15648   1.399920   0.622831  -1.860259  4.095776  ... -5.742194 -2.019131   
165842 -1.017692   0.237848  -0.172001 -0.489069  ...  0.069270  0.275828   
15451  -4.384491 -15.939003  13.696416 -3.948455  ...  1.592754  1.754608   

             V22       V23       V24       V25       V26       V27       V28  \
247995  0.445744 -0.141136 -0.265517  0.362260 -0.416062  0.507370  0.243744   
189888  0.883051 -0.597059  0.952351  1.150571 -0.789611 -0.307360  0.391878   
15648   1.896375  1.359321  0.043460  0.683422  0.791735 -2.854021  1.168464   
165842  0.836055 -0.274084 -0.058501  0.025833 -0.203120 -0.026987  0.299542   
15451  -1.466115 -0.856779  0.125777  1.402587 -0.223755  1.574249  0.469201   

        Amount  
247995   51.37  
189888   16.99  
15648    88.00  
165842   39.99  
15451    99.99  

[5 rows x 30 columns]

#### Training the model on Logistic Regression

In [40]:
model = LR()

In [42]:
model.fit(X_train,Y_train)

LogisticRegression()

### 5. Model Evaluation 

Using Accuracy score to determine the accuracy of the training model 

In [55]:
training_prediction=model.predict(X_train)
print(training_prediction)

[0 0 0 0 0 0 0 1 0 1 0 1 0 1 1 0 0 1 1 0 0 1 0 0 0 1 0 1 0 1 1 0 0 0 1 0 1
 0 0 0 0 0 1 1 1 0 0 0 1 0 1 0 1 0 0 0 0 1 1 0 0 0 1 0 1 0 1 1 0 0 1 1 1 0
 0 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 1 1 0 0 1 0 0 0 0 1
 1 1 0 0 0 1 0 1 0 1 0 1 0 0 1 1 1 0 0 0 1 1 1 1 1 0 1 0 1 0 0 0 0 1 1 0 0
 1 0 0 1 1 1 0 0 0 1 0 0 0 1 0 0 1 1 1 0 1 0 0 0 1 0 1 1 0 0 0 1 1 0 0 0 0
 0 0 0 1 0 1 0 1 0 0 0 0 1 0 1 0 0 0 1 1 1 1 1 1 0 0 1 1 1 1 0 1 1 1 0 1 1
 0 1 0 1 0 0 1 0 0 0 0 1 1 1 0 0 0 1 0 1 0 1 1 1 0 0 0 0 0 1 0 1 1 1 0 0 0
 1 0 0 1 0 0 0 1 1 1 1 1 1 1 1 0 1 0 0 0 1 0 1 1 0 0 1 0 1 0 0 0 1 1 0 0 1
 1 1 1 0 1 1 0 1 0 1 0 1 1 1 0 0 0 1 0 1 1 0 1 0 1 1 1 1 0 1 0 0 0 0 1 0 0
 0 0 0 0 0 1 1 0 1 1 0 1 0 0 0 0 1 0 1 1 0 1 1 0 1 1 1 1 0 1 1 0 0 1 0 1 0
 0 0 0 1 0 1 0 0 1 1 0 0 0 0 1 1 0 1 1 1 1 1 1 1 0 0 0 0 1 0 0 0 0 0 0 0 0
 0 1 0 0 0 0 0 0 0 0 1 0 1 1 0 0 0 0 0 0 1 0 1 0 1 0 0 1 0 1 1 0 0 1 0 0 1
 1 0 1 0 0 1 0 1 0 1 0 1 1 1 0 1 0 1 0 1 0 0 0 0 0 1 0 1 1 0 1 0 1 1 0 1 0
 0 0 0 1 1 0 1 0 1 1 0 0 

In [52]:
training_accuracy = accuracy_score(Y_train,training_prediction)
print(f"The training accuracy is {training_accuracy}")

The training accuracy is 0.9034307496823379


In [56]:
testing_prediction = model.predict(X_test)
print(testing_prediction)

[1 0 0 0 1 0 1 1 0 1 0 1 1 0 0 1 0 0 0 0 0 0 0 1 0 1 1 0 1 0 0 1 0 1 1 0 0
 0 0 1 1 0 0 1 1 1 1 1 0 0 0 0 0 0 0 0 0 1 0 0 1 1 1 1 1 0 1 0 0 1 0 0 0 0
 1 1 0 1 1 1 0 1 1 1 0 0 0 1 0 0 0 1 1 1 1 1 1 1 0 1 1 1 1 0 1 0 0 0 0 0 0
 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0 0 1 0 1 1 0 0 1 0 1 0 1 0 1 1 0 1 1 0 0 1 0
 1 0 0 0 1 1 0 0 1 1 0 0 0 1 0 1 1 0 0 1 0 1 1 0 1 1 0 0 0 0 0 0 0 1 0 0 1
 0 0 0 0 0 1 1 0 1 1 1 0]


In [57]:
testing_accuracy = accuracy_score(Y_test,testing_prediction)
print(f"The testing accuracy is {testing_accuracy}")

The testing accuracy is 0.9086294416243654


Our model shows accuracy of 90.863% on testing data